moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,299 @@
+ */
+
+#include <moveit/online_signal_smoothing/acceleration_filter.h>
+#include <rclcpp/logging.hpp>
+#include <Eigen/Sparse>
+
+// Disable -Wold-style-cast because all _THROTTLE macros trigger this
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+
+namespace online_signal_smoothing
+{
+// The threshold below which any velocity or position difference is considered zero.
+constexpr double COMMAND_DIFFERENCE_THRESHOLD = 1E-4;
+constexpr double ALPHA_UPPER_BOUND = 1.0;
+constexpr double ALPHA_LOWER_BOUND = 0.0;
+
+struct CSCWrapper
+{
+  std::vector<c_int> i;
+  std::vector<c_int> j;
+  std::vector<double> x;
+  csc sparse_matrix;
+
+  CSCWrapper(Eigen::SparseMatrix<double>& M)
+  {
+    M.makeCompressed();
+
+    sparse_matrix.n = M.cols();
+    sparse_matrix.m = M.rows();
+    i.assign(M.innerSize(), 0);
+    sparse_matrix.i = i.data();
+    j.assign(M.outerSize() + 1, 0);
+    sparse_matrix.p = j.data();
+    sparse_matrix.nzmax = M.nonZeros();
+    sparse_matrix.nz = -1;
+    x.assign(M.nonZeros(), 0.0);
+    sparse_matrix.x = x.data();
+
+    update(M);
+  }
+
+  void update(Eigen::SparseMatrix<double>& M)
+  {
+    for (size_t ind = 0; ind < i.size(); ++ind)
+    {
+      i[ind] = M.innerIndexPtr()[ind];
+    }
+
+    for (size_t ind = 0; ind < j.size(); ++ind)
+    {
+      j[ind] = M.outerIndexPtr()[ind];
+    }
+    for (size_t ind = 0; ind < x.size(); ++ind)
+    {
+      x[ind] = M.data().at(ind);
+    }
+  }
+};
+
+MOVEIT_STRUCT_FORWARD(OSQPDataWrapper);
+
+struct OSQPDataWrapper
+{
+  OSQPDataWrapper(Eigen::SparseMatrix<double>& P_sparse, Eigen::SparseMatrix<double>& A_sparse)
+    : P{ P_sparse }, A{ A_sparse }
+  {
+    data.n = P_sparse.rows();
+    data.m = A_sparse.rows();
+    data.P = &P.sparse_matrix;
+    q = Eigen::VectorXd::Zero(P_sparse.rows());
+    data.q = q.data();
+    data.A = &A.sparse_matrix;
+    l = Eigen::VectorXd::Zero(A_sparse.rows());
+    data.l = l.data();
+    u = Eigen::VectorXd::Zero(A_sparse.rows());
+    data.u = u.data();
+  }
+
+  void update_A(OSQPWorkspace* work, Eigen::SparseMatrix<double>& A_sparse)
+  {
+    A_sparse.makeCompressed();
+    A.update(A_sparse);
+    osqp_update_A(work, A.x.data(), OSQP_NULL, A.x.size());
+  }
+
+  CSCWrapper P;
+  CSCWrapper A;
+  Eigen::VectorXd q;
+  Eigen::VectorXd l;
+  Eigen::VectorXd u;
+  OSQPData data{};
+};
+bool AccelerationLimitedPlugin::initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model,
+                                           size_t num_joints)
+{
+  // copy inputs into member variables
+  node_ = node;
+  num_joints_ = num_joints;
+  robot_model_ = robot_model;
+  cur_acceleration_ = Eigen::VectorXd::Zero(num_joints);
+
+  // get node parameters and store in member variables
+  auto param_listener = online_signal_smoothing::ParamListener(node_);
+  auto params = param_listener.get_params();
+  update_rate_ = params.update_rate;
```

Comment:
I like storing the params object directly and use `params_,MY_PARAM` everywhere else. I think it helps readability and removes a bunch of individual variables. You could consider doing that too, if you don't like it, keep it as it is :+1:

Reply:
Okay, I removed the other variables from the class